backend/apid/actions/roles.go

```diff
@@ -82,23 +81,17 @@ func (a RoleController) Get(ctx context.Context, name string) (*types.Role, erro
 }
 
 // List returns all available roles.
-func (a RoleController) List(ctx context.Context) ([]*types.Role, string, error) {
-	pageSize := corev2.PageSizeFromContext(ctx)
-	continueToken := corev2.PageContinueFromContext(ctx)
-
+func (a RoleController) List(ctx context.Context, pred *store.SelectionPredicate) ([]corev2.Resource, error) {
 	// Fetch from store
-	results, newContinueToken, err := a.Store.ListRoles(ctx, int64(pageSize), continueToken)
+	results, err := a.store.ListRoles(ctx, pred)
 	if err != nil {
-		switch err := err.(type) {
-		case *store.ErrNotFound:
-			return nil, "", NewErrorf(NotFound)
-		default:
-			return nil, "", NewError(InternalErr, err)
-		}
+		return nil, NewError(InternalErr, err)
 	}
 
-	// Encode the continue token with base64url (RFC 4648), without padding
-	encodedNewContinueToken := base64.RawURLEncoding.EncodeToString([]byte(newContinueToken))
+	resources := make([]corev2.Resource, len(results))
```

Comment:
I see this code block is repeated often in the refactor. Is this something we could maybe make a generic function?

Reply:
At first I tried to create a function like this, but obviously it won't work because an `interface{}` doesn't implement the corev2.Resource interface:

```go
func returnResources(results []corev2.Resource) []corev2.Resource {
	resources := make([]corev2.Resource, len(results))
	for i, v := range results {
		resources[i] = corev2.Resource(v)
	}
	return resources
}
```

If someone has an idea, I'm all ears!

Reply:
Unfortunately, in Go, even though `[]corev2.Role` would hand-wavely qualify as a `[]corev2.Resource` because `Role` meets the `Resource` interface, the compiler won't accept it because it would have to construct a potentially expensive operation behind the scene to do what you want: convert each `corev2.Role` into a `Resource` interface (they have different memory representations), which is O(n). This is what you explicitly do with that `for` loop. If the compiler did that "magic", you could directly use `results` (type `[]corev2.Role`) as the function's return value (which needs to be `[]corev2.Resource`); no need for that for loop!

So in Go, boilerplate it is! If you want to abstract it away you can still write that `[]interface{} -> []corev2.Resource` function provided you're willing to use reflection. Something like that should do the trick:

```go
func returnResources(input interface{}) []corev2.Resource {
	var resources []corev2.Resource

	v := reflect.ValueOf(input)

	if v.Kind() != reflect.Slice {
		panic("input needs to be a slice")
	}

	for i := 0; i < v.Len(); i++ {
		resource := v.Index(i).Interface().(corev2.Resource)
		resources = append(resources, resource)
	}

	return resources
}
```

Note that the type assertion will panic if the elements of the input slice do not meet the `corev2.Resource` interface.

Reply:
This is purely aspirational, but I appreciate that you tried to simplify it in your approach @palourde! While @ccressent approach would likely work, it may not necessary to introduce reflection for a small win over code duplication.